aesd-char-driver/main.c

```diff
@@ -0,0 +1,271 @@
+/**
+ * @file aesdchar.c
+ * @brief Functions and data related to the AESD char driver implementation
+ *
+ * Based on the implementation of the "scull" device driver, found in
+ * Linux Device Drivers example code.
+ *
+ * @author Dan Walkes
+ * @date 2019-10-22
+ * @copyright Copyright (c) 2019
+ *
+ */
+
+#include <linux/module.h>
+#include <linux/init.h>
+#include <linux/printk.h>
+#include <linux/types.h>
+#include <linux/cdev.h>
+#include <linux/fs.h> // file_operations
+#include "aesdchar.h"
+#include "linux/slab.h"
+#include "linux/string.h"
+int aesd_major =   0; // use dynamic major
+int aesd_minor =   0;
+
+MODULE_AUTHOR("MaanasMDK"); /** TODO: fill in your name **/
+MODULE_LICENSE("Dual BSD/GPL");
+
+struct aesd_dev aesd_device;
+
+int aesd_open(struct inode *inode, struct file *filp)
+{
+    struct aesd_dev *dev;
+    PDEBUG("open");
+    /**
+     * TODO: handle open
+     */
+    dev = container_of(inode->i_cdev, struct aesd_dev, cdev);
+    filp->private_data = dev;
+    return 0;
+}
+
+int aesd_release(struct inode *inode, struct file *filp)
+{
+    PDEBUG("release");
+    /**
+     * TODO: handle release
+     */
+    return 0;
+}
+
+ssize_t aesd_read(struct file *filp, char __user *buf, size_t count,
+                loff_t *f_pos)
+{
+    ssize_t retval = 0;
+    struct aesd_dev *dev;
+    struct aesd_buffer_entry *tmp_buf;
+    int tmp_buf_count = 0;
+    size_t offset_bytes;
+    dev = filp->private_data;
+    PDEBUG("read %zu bytes with offset %lld",count,*f_pos);
+    /**
+     * TODO: handle read
+     */
+
+    // lock the read access
+    mutex_lock(&aesd_device.lock);
+
+    tmp_buf = aesd_circular_buffer_find_entry_offset_for_fpos(&dev->buffer, *f_pos, &offset_bytes);
+    // check if count doesn't exist
+    if( tmp_buf == NULL )
+    {
+        *f_pos = 0;
+        goto cleanup;
+    }
+
+    // check if tmp_buf has required count
+    if( (tmp_buf->size - offset_bytes) < count )
+    {
+        *f_pos += tmp_buf->size - offset_bytes;
+        tmp_buf_count = tmp_buf->size - offset_bytes;
+    }
+    else
+    {
+        *f_pos += count;
+        tmp_buf_count = count;
+    }
+
+    // copy to user buffer
+    if( copy_to_user(buf, tmp_buf->buffptr+offset_bytes, tmp_buf_count))
+    {
+        retval = -EFAULT;
+        goto cleanup;
+    }
+
+    retval = tmp_buf_count;
+
+    cleanup : mutex_unlock(&aesd_device.lock);
+    
+    return retval;
+}
+
+ssize_t aesd_write(struct file *filp, const char __user *buf, size_t count,
+                loff_t *f_pos)
+{
+    char *tmp_buf;
+    bool packet_flag = false;
+    struct aesd_dev *dev;
+    int packet_len = 0;
+    struct aesd_buffer_entry write_entry;
+    char *ret;
+    int i;
+    ssize_t retval = 0;
+    PDEBUG("write %zu bytes with offset %lld",count,*f_pos);
+    /**
+     * TODO: handle write
+     */
+    dev = filp->private_data;
+
+    mutex_lock(&aesd_device.lock);
+
+    // malloc a new buffer
+    tmp_buf = (char *)kmalloc(count, GFP_KERNEL);
```

Comment:
@MaanasMDK does this need to be inside the locking interval?

Reply:
Not necessarily. My idea was to keep the entire driver functionality appear atomic but since it's a temporary buffer I could keep it outside the locking interval.